keps/sig-api-machinery/2876-crd-validation-expression-language/README.md

```diff
@@ -348,25 +349,104 @@ to fail with a descriptive error.
+
+- `size` (or string, lists, maps, ...)
+- String inspection: `startsWith`, `endsWith`, `contains`
+- Regex: `matches`
+- Date/time accessors: `getDate`, `getDayOfMonth`, `getDayOfWeek`, `getDayOfYear`, `getFullYear`, `getHours`, `getMilliseconds`, `getMinutes`, `getMonth`, `getSeconds`
+
+CEL standard [macros](https://github.com/google/cel-spec/blob/master/doc/langdef.md#macros) include:
+
+- `has`
+- `all`
+- `exists`
+- `exists_one`
+- `map`
+- `filter`
+
+CEL [extended string function library](https://github.com/google/cel-go/blob/master/ext/strings.go) includes:
+- 
+- `charAt`
+- `indexOf`, `lastIndexOf`
+- `upperAscii`, `lowerAscii`
+- `replace` (string literal replacement, not regex)
+- `split`
+- `substring`
+- `trim`
+
+Notable absences from the standard functions, when comparing against the core libraries of Go, Java and Python are:
+
+Strings:
+
+- trim/trimLeft/trimRight (trimming either by whitespace or by provided cutset)
+- trimPrefix/trimSuffix
+- join
+- replace (regex)
+- format (e.g. `fmt.Sprintf()`)
+
+Lists:
+
+- sort
+- indexOf/lastIndexOf
+- replace
+- sublist
+- split
+- reverse
+- reduce or sum/min/max (reduce is more general, the others are more concrete)
+
+Numbers:
+
+- abs/cel/floor/round
+- exp/log/log10/logN/pow/sqrt
+- max(x, y, ..)/min(x, y, ...)
+- <trig functions>
+
+Maps:
+
+- <Ability to construct a map using a comprehension>
+
+Some considerations when selecting which of the above we should include in CEL expressions in Kubernetes:
+
+- We should include functions that will be needed for future use cases of CEL in Kubernetes (e.g. general admission) as
+  ones needed for validation.
+- Regex replace is very powerful and useful. It is also potentially dangerous due to its ability to allocate memory.
+- `format`: Since CEL supports string concatenation, the value of having format would only be to do things like format floats.
+  Formatting functions are complex and require extensive documentation to teach.
+- `sort` makes it possible to check if a list is sorted. It can be an expensive operation.
+- `indexOf` / `lastIndexOf` / `split` / `replace` (the string functions) could be overloaded to provide the equivalent list functions?
+   Keeping string and list functions consistent seems like a good way to keep the learning curve down.
+- `reduce` is going to be non-obvious to anyone that doesn't have a functional programming background?
+  - If we provide `reduce` are developers going to also expect to have `fold` or `zip`?
+- "Ability to construct a map using a comprehension": this appears mechanically problematic to support in CEL?
+
+Proposal:
+
+- Keep all the CEL standard functions and macros as well as the extended string library.
+- Add regex replace support. It's very useful, and we already allow regex matching.
+- Add `trim` / `trimLeft` / `trimRight` (overloaded to take an optional cutset arg) for strings
+- Add `trimPrefix` / `trimSuffix` for strings
```

Comment:
The use cases that immediately come to mind for trimming characters from the start or end of a string are ones I would not recommend (e.g. tolerating optional prefix/suffix whitespace, rather than making validation requirements strict). Do you have examples of other uses?

Reply:
I agree. The only use case I've hit in the past was:

- an int-or-string is set to a percentage, e.g. "80%"
- The validation step wants to ensure the percentage is valid

One naive way to do this is trim the '%' suffix and then convert the remaining string to a number perform whatever numeric checks you want to that number.

I'll admit that regex matching can do everything needed here in most use cases. The only exception might be if the numeric checks are non-trivial, in which case extracting out the number provides some value.

Reply:
Dropped this one.